groups/nts/ntsa/ntsa_tcpcongestioncontrol.h

```diff
@@ -0,0 +1,193 @@
+// Copyright 2024 Bloomberg Finance L.P.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef INCLUDED_NTSA_TCPCONGESTIONCONTROL
+#define INCLUDED_NTSA_TCPCONGESTIONCONTROL
+
+#include <bsls_ident.h>
+BSLS_IDENT("$Id: $")
+
+#include <ntsa_error.h>
+#include <ntscfg_platform.h>
+#include <ntsscm_version.h>
+#include <bsl_array.h>
+#include <bsl_ostream.h>
+#include <bsl_string.h>
+#include <bsl_span.h>
+
+namespace BloombergLP {
+namespace ntsa {
+
+/// Provide a value identifying tcp congestion control algorithm.
+///
+/// @details
+/// Provide a value-semantic type that identifies tcp congestion control
+/// algorithm.
+///
+/// @par Attributes
+/// This class is composed of the following attributes.
+///
+/// @li d_buffer:
+/// Buffer which contains name of the tcp congestion control algorithm.
+///
+/// @par Thread Safety
+/// This class is not thread safe.
+///
+/// @ingroup module_ntsa_system
+
+class TcpCongestionControl
+{
+  public:
+    enum { TCP_CA_NAME_MAX = 16 };
+
+    enum Algorithm {
+        e_RENO = 0,
+        e_CUBIC,
+        e_BBR,
+        e_BIC,
+        e_DCTCP,
+        e_DIAG,
+        e_HIGHSPEED,
+        e_HTCP,
+        e_HYBLA,
+        e_ILLINOIS,
+        e_LP,
+        e_NV,
+        e_SCALABLE,
+        e_VEGAS,
+        e_VENO,
+        e_WESTWOOD,
+        e_YEAH
+    };
+
+  private:
+    enum { BUFFER_SIZE = TCP_CA_NAME_MAX + 1 };
+
+    bsl::array<char, BUFFER_SIZE> d_buffer;
```

Comment:
I would prefer to simply make the representation of the algorithm name a `bsl::string`, with its associated manipulator member function `setAlgorithm(const bsl::string_view& value)` and associated accessor member function `const bsl::string& algorithm()`. Let's keep this type free of platform-specific knowledge. I think it's unnecessarily confusing to have a span<char> that must include a null terminator, as a generic representation of this state. Let's move the non-portable details of how we set or get the algorithm, with platform-specific requirements of null-terminator and sizes, to `ntsu::SocketOptionUtil`. We don't have any efficiency concern here because getting and setting the congestion control algorithm can't really appear on any fast path, so let's favor clarity (and note the situation here is different than ntsa::LocalName, which does appear on the fast, typical path.)

Reply:
Ok, but it means that this type will become allocator aware (and ntsa::SocketIOption also)

Reply:
Done